WordPress/gutenberg
# [RNMobile] Introduce Context Constants to Native

`Gutenberg Mobile:` https://github.com/wordpress-mobile/gutenberg-mobile/pull/4099

## Description

The Gallery block was [recently refactored](https://github.com/WordPress/gutenberg/pull/25940), for both the web and the mobile (native) apps, in order to pull in image blocks via [the InnerBlocks API](https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/nested-blocks-inner-blocks/). 

As part of this refactor for the mobile apps, the following native-specific [block contexts](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-context/) were introduced in order to style the gallery's inner image blocks:
* `imageCrop`, was introduced to support the cropping of images. 
* `fixedHeight`, was introduced in https://github.com/WordPress/gutenberg/pull/36013 to explicitly set a fixed height for the images.

Similarly, as described in https://github.com/WordPress/gutenberg/issues/33531, an `allowResize` context was introduced on the web-side to disable the image block's resize option for galleries. 

<hr />

**Problem:** Additional contexts for customizing child blocks are going to be necessary. 

On the native side, for example, the settings icon is included within the block, unlike on the web, where the block's settings are separated out into the sidebar. The ability to hide the settings icon is therefore likely to be a fairly common use case for parent blocks. There are also likely to be other valid use cases for customizing inner blocks in the future.

Adding additional contexts isn't an ideal or scaleable solution. This has also been discussed in https://github.com/WordPress/gutenberg/issues/33531, with some suggestions outlined there.

<hr />

**Proposal:** With this PR, the idea of introducing context constants, as initially [suggested here](https://github.com/WordPress/gutenberg/pull/25940#discussion_r671011374), is built upon. 

`usesContextConstants` is provided as an alternative to `usesContext` and `providesContextConstants ` as an alternative to `providesContext`.

The aim is to explicitly indicate the purpose of these contexts and that they aren't serialised, as would normally be the case with contexts. It is also hoped that the introduction of context constants will make it easier to add these types of contexts in the future.

<hr />

**_Note: This PR is native-specific but a separate web-specific PR can be made, building on the work done here, if approved._**

## How has this been tested?

_Prequisite: Enable the flag for the refactored Gallery block. on mobile. At the time of writing, the way to enable the block on the Gutenberg Mobile demo app is to comment out [this section of code](https://github.com/WordPress/gutenberg/blob/5c9e84d66419c75900b8e6c5addc6eb2326e203e/packages/block-library/src/gallery/edit-wrapper.js#L33-L44)._ 

With the refactored Gallery block enabled, we should ensure that the changes introduced with this PR don't have any adverse effects:

* Navigate to the editor within either the Android or iOS app and add the Gallery block.
* Add a selection of images to the block. Verify that the images appear, which will in turn verify that the new `fixedHeight` context constant is working and hasn't caused any regressions (no images would be visible if the context wasn't working).
* Select the cog/gear icon to access the block's settings. Toggle the "crop images" tab on and off to verify that the images within the gallery change from being cropped, to not. This will confirm that the `imageCrop` context is still working as expected.

## Screenshots 

The following screencast shows the images going from cropped to uncropped, as expected, to show the `imageCrop` context is working as intended following the changes in this PR:

https://user-images.githubusercontent.com/2998162/136827580-28d8e53a-4f90-42fc-bd37-ccb60c6c8514.mov

## Types of changes

This PR includes changes to code and the overall approach that's taken to adding contexts when needed to change inner blocks on the native-side, with the following high-level overview:

* `providesContextConstants` and `usesContextConstants` are added to block metadata upon registration [via the `blocks/src/api/registration.js` file](https://github.com/WordPress/gutenberg/pull/35510/files#diff-30ce126b90014d659302087765876898f16562219f796796ed407fdb346c8e4d). 
* `providesContextConstants` was [added to the object generated](https://github.com/WordPress/gutenberg/pull/35510/files#diff-f7f5a30681413bf9eb7368f36c36184b454dd2094c49c4c21930d414884d8624R35) by the `getBlockContext` function. There are also now checks in place [here](https://github.com/WordPress/gutenberg/pull/35510/files#diff-22dc9560179184651a38572ecad3bef26c968be9677771ddacf13fc8c131795bR34) and [here](https://github.com/WordPress/gutenberg/pull/35510/files#diff-b2f4aa1258856994e49ed89c4068795b171a9b7b7b90cc1440c272517b1792ecR140) to account for the existence of context constants on the native side.
* The Gallery and Image blocks have been refactored to use `fixedHeight` (an existing context, first implemented in https://github.com/WordPress/gutenberg/pull/36013) as an example of a context constant.

## Checklist:
